manuskript/ui/editors/editorWidget.py

```diff
@@ -293,9 +293,8 @@ def addSpacer():
 
         try:
             self._model.dataChanged.connect(self.modelDataChanged, AUC)
-            self._model.rowsInserted.connect(self.updateIndexFromID, AUC)
-            self._model.rowsRemoved.connect(self.updateIndexFromID, AUC)
-            #self.mw.mdlOutline.rowsAboutToBeRemoved.connect(self.rowsAboutToBeRemoved, AUC)
+            self._model.rowsInserted.connect(self.rowsInserted, AUC)
```

Comment:
Did you check to see if the index needs to change when a row is inserted or if the old index remains valid since the outlineItem itself doesn't change?

Reply:
My suspicion is that it remains valid. I cannot reproduce buggy behaviour. But do we want to count on that?

From what I can tell, `model.createIndex()` relies on a row, column and item to create said index. Looking at the manual, it exposes `row()` and `column()` functions for accessing this data. The class also appears as very lightweight in its design and does not receive any updates. Hell, I just spotted this in the manual:

> **Note**: Model indexes should be used immediately and then discarded. You should not rely on indexes to remain valid after calling model functions that change the structure of the model or delete items. If you need to keep a model index over time use a QPersistentModelIndex.

This leads me to believe that inserting an item before a currently stored index would indeed cause the QModelIndex to be invalid. Which makes me wonder: should we not be using `QPersistentModelIndex` instead? About that class, the manual tells us:

> A QPersistentModelIndex is a model index that can be stored by an application, and later used to access information in a model. Unlike the QModelIndex class, it is safe to store a QPersistentModelIndex since the model will ensure that references to items will continue to be valid as long as they can be accessed by the model.

I am not going to make any such change in this patch. The current infrastructure that receives signals is adequate to monitor such changes and apply them where necessary.

That said, I did just notice it fails to update the tab title when we rename an open item. I guess that fix now gets to piggyback along with this patch. 😁

Reply:
Ah, that's really interesting. I guess we would want to switch to a QPersistentModelIndex instead of a QModelIndex (not in this patch though). And thanks, that does answer my question about the need for rowsInserted.